---
title: "visitation_bias"
author: "Doug Sponsler"
date: "12/14/2020"
output: html_document
---
# Preparations
## Packages 
```{r, warning=FALSE, include=FALSE, echo = FALSE}
library(bipartite)
library(mgcv)
library(mgcViz)
library(brms)
library(lme4)
#library(inla)
library(bayesplot)
library(vegan)
#library(tidybayes)
library(tidyverse)
library(lubridate)
library(tidymodels)
```

## Load data
```{r message=FALSE, warning=FALSE, include=FALSE, echo = FALSE}
# Site data
site_data <- read_csv("../processed_data/site_data.csv") %>%
  dplyr::select(-c(slope.calc, slope.est, elev.min, elev.max, temp.mean)) %>% # drop these variables
  mutate(elev.class2 = factor(elev.class2, levels = c("low", "mid", "high")),
         site = factor(site)) # turn elav.class into an ordered factor and site into a regular factor

# BB traits
bb_traits <- read_csv("../processed_data/bb_traits.csv") 

# FL traits
## family-level taxonomy
fl_tax <- read_csv("../processed_data/floral_tax.csv") 

## Kugler morphotypes (accessed through Bioflor)
fl_ktype <- read_csv("../processed_data/floral_k_type.csv") 

## Flower colors (accessed through BioFlor)
fl_color <- read_csv("../processed_data/floral_color.csv")

## diameter and corolla depth data from Gita Benadi
fl_morph <- read_csv("../processed_data/flower_morphology_GitaBenadi.csv") %>% 
  dplyr::select(SampleID, plant.family = Family, plant.sp = Species, 
                Unit, Flowers_per_unit, diam = Diameter, LN) %>%
  group_by(plant.sp) %>%
  summarize(diam = mean(diam), LN = mean(LN))

## verbal description of Kugler morphotypes
fl_ktype_key <- read_delim("../processed_data/kugler_key.tsv", delim = "\t", col_names = FALSE) %>%
  dplyr::select(k.type = 1, description = 2)

## join all floral trait tables
fl_traits <- fl_tax %>%
  full_join(fl_ktype) %>%
  full_join(fl_color) %>%
  full_join(fl_morph) %>%
  dplyr::select(plant.sp, plant.genus, plant.family, k.type, k.type.s, k.type.ss, color, diam, LN)
```

## Align survey and visitation data with pseudodates
```{r}
# Site-dates where BB visitation was recorded (excluding dates where only males were recorded)
visitation_samples <- read_csv("../processed_data/network.csv") %>%
  filter(caste != "male") %>%
  select(site, date) %>%
  distinct() 

# Site dates where floral surveying was conducted
survey_samples <- read_csv("../processed_data/floral_survey.csv") %>%
  select(site, date) %>%
  distinct()

# Site-dates where visitation sampling and floral sampling coincided
shared_dates <- inner_join(visitation_samples, survey_samples)

# Site-dates where visitation was recorded WITHOUT corresponding survey data
visitation_only <- anti_join(visitation_samples, survey_samples) %>%
  mutate(date.v = date) %>%
  select(-date)

# This creates a pseudo date column for unmatched visitation site-dates that will match each visitation date
# to its nearest survey data. For site t4, there is a 2-year mismatch; we will drop these site-dates later.
pseudodates <- full_join(visitation_only, survey_samples) %>%
  mutate(mismatch = abs(date.v - date)) %>%
  group_by(site, date.v) %>%
  filter(mismatch == min(mismatch)) %>%
  rename(date = date.v, pseudo.date = date)
```

## Generate a data frame that represents visitation and survey abundance
```{r}
# BB visitation data
visitation <- read_csv("../processed_data/network.csv") %>%
  filter(caste != "male") %>% # we will ignore males 
  group_by(site, date, bb.sp, plant.sp) %>%
  summarize(visits = n()) %>% # visits per BB-FL pair
  ungroup() %>%
  left_join(pseudodates) %>% # Now we will replace unmatched dates with their pseudo dates
  replace_na(replace = list(mismatch = 0)) %>% # set mismatch to zero for perfectly aligned dates
  filter(mismatch <= 7) %>% # this gets rid of irreconcilable dates (like the t4 dates mentioned above)
  mutate(date2 = if_else(is.na(pseudo.date), date, pseudo.date)) %>% # replace date with pseudodate for all site-dates lacking perfect alignments
  select(-c(date, pseudo.date, mismatch)) %>%
  rename(date = date2) %>%
  group_by(site, date) %>%
  complete(bb.sp, plant.sp, fill = list(visits = 0)) # generate all possible BB-FL combos for each site date and fill in zeros for visits

# Floral cover data
flower_cover <- read_csv("../processed_data/floral_survey.csv") %>%
  filter(flower.cover > 0) %>% # ignore plants that were not blooming
  group_by(site, date, plant.sp) %>%
  summarize(flower.cover = sum(flower.cover))

# Get shared dates after adding pseudodates. This will let us filter out survey dates that don't
# have corresponding visitation sampling dates
shared_dates_with_pseudo <- inner_join(visitation, flower_cover, by = c("site", "date")) %>%
  select(site, date) %>%
  distinct()

# Join visitation and flower cover data

## First we need a list of all BB spp. by site and date
sp_by_site <- visitation %>% 
  ungroup() %>%
  select(site, date, bb.sp) %>%
  distinct()

## Then we need a list of unvisited plant.sp by site and date
unvisited <- flower_cover %>% 
  semi_join(shared_dates_with_pseudo) %>% # filter flower cover data to include only site-dates with visitation complement (real or pseudodates)
  full_join(visitation) %>% # join to visitation data
  filter(is.na(bb.sp)) %>% # select only the site-date-plants that were totally unvisited (and therefore have no BB spp to join to)
  select(-bb.sp) %>% # drop bb.sp variable because I just want the plants
  left_join(sp_by_site) %>% # join in the species list by site and date, thus populating the unvisited plants with all the BB spp recorded for each site-date
  replace_na(replace = list(visits = 0)) # replace NAs (created in the line above) with zero

## Now we can combine everything into a single data frame
visit_cover <- flower_cover %>% 
  semi_join(shared_dates_with_pseudo) %>% # filter flower cover data to include only site-dates with visitation complement (real or pseudodates)
  full_join(visitation) %>% # join to visitation data
  replace_na(replace = list(flower.cover = 0.01)) %>% # If a plant was visited but not surveyed, we will assign it the minimum non-zero flower cover value of 0.01.
  filter(!is.na(bb.sp)) %>% # drop the univisited plants
  bind_rows(unvisited) %>% # add them back, with BB spp and zeros
  select(site, date, bb.sp, plant.sp, visits, flower.cover) %>%
  group_by(site, date, bb.sp, plant.sp, flower.cover) %>% # In a few cases, the use of pseudodates causes two sampling period to converge on the same date, which creates apparent duplicate visitation data. We will handle this by averaging the visits.
  summarize(visits = mean(visits)) %>%
  left_join(fl_traits) %>% # add floral traits
  ungroup() %>%
  mutate(k.type.ss = factor(k.type.ss))

## Get total visits by BB sp and site-date; this will be a measure of BB abundance that we can use after we've calculated d'. I could, of course, work this into the biaseval function, but I'll put that off for now.
bb_abundance <- visit_cover %>%
  group_by(site, date, bb.sp) %>%
  summarize(bb.sp.abund = sum(visits)) %>%
  group_by(site, date) %>%
  mutate(bb.sp.prop.abund = bb.sp.abund/sum(bb.sp.abund))

fl_abundance <- visit_cover %>%
  group_by(site, date) %>%
  summarize(total.flower.cover = sum(flower.cover))
```

## A handy function for calculating Blüthgen's d'
```{r}
# A function for calculating Blüthgen's d', parsed in various ways
biaseval <- function(x, criterion, ...) { # criterion = plant grouping variable, e.g. plant.sp, k.type.ss, color
  
  criterion <- enquo(criterion) # tidy eval
  
  x %>%
    
# Select variables of interest; the ... represents the variables we will use to parse our analysis
    select(..., bb.sp, !!criterion, flower.cover, visits) %>% 
    group_by(..., bb.sp, !!criterion) %>%
    
# If criterion = plant.sp, this next step does nothing. But if criterion is some other grouping variable (e.g. k.type.ss), this sums flower cover and visits over that grouping variable.
    summarize(flower.cover = sum(flower.cover), 
              visits = sum(visits)) %>%
    
# Nest into list column
    group_by(...) %>%
    nest() %>%

# Pivot each nested data frame into the kind of web that bipartite likes
    mutate(webs = map(data, function(x) select(x, bb.sp, !!criterion, visits) %>%
                        pivot_wider(names_from = !!criterion, values_from = visits, values_fill = 0) %>%
                        column_to_rownames("bb.sp"))) %>%

# Pull out the flower.cover variable to use in the "abuns" argument of dfun below
    mutate(low.abund = map(data, function(x) select(x, !!criterion, flower.cover) %>%
                            unique() %>%
                            column_to_rownames(paste(criterion)[2]))) %>% # Tell me why I need to do this, Hadley.

# Call the dfun command from bipartite to calculate Blüthgen's d'
    mutate(d = map2(webs, low.abund, function(x, y) bipartite::dfun(x, abuns = y$flower.cover) %>%
                      data.frame() %>%
                      rownames_to_column(var = "bb.sp"))) %>%
    
    unnest(cols = d) %>%
    na.omit() %>% # in a few cases where the sampling is very sparse, d' cannot be calculated; these are safe to drop
    ungroup() %>%
    mutate(site = factor(site),
         bb.sp = factor(bb.sp),
         year = factor(lubridate::year(date)))
}
```

# 1. Is bumble bee foraging biased? How do taxonomic, phylogenetic, and morphological bias compare -- NO, can't analyze this, these should be considered separate repsonse variables and analyzed in separate models? Are some BB species more biased than others?
## 1.1 Calculate Blüthgen's d' (i.e. visitation bias)
```{r}
# No grouping -- just plant.sp
bias_plantsp <- biaseval(visit_cover, plant.sp, site, date) %>%
  left_join(site_data) %>% # add site data
  left_join(bb_abundance) %>% # add BB abundance data
  left_join(fl_abundance) %>% # add fl abundance
  mutate(criterion = factor(rep("plant.sp", n())))

# Grouped by k.type.ss
bias_ktype <- biaseval(visit_cover, k.type.ss, site, date) %>%
  left_join(site_data) %>%
  left_join(bb_abundance) %>% # add BB abundance data
  left_join(fl_abundance) %>% # add fl abundance
  mutate(criterion = factor(rep("ktype", n()))) 

# Grouped by plant.family
bias_family <- biaseval(visit_cover, plant.family, site, date) %>%
  left_join(site_data) %>%
  left_join(bb_abundance) %>% # add BB abundance data
  left_join(fl_abundance) %>% # add fl abundance
  mutate(criterion = factor(rep("plant.family", n())))

# Collated
bias <- bind_rows(bias_plantsp, 
                  bias_ktype, 
                  bias_family) %>%
  mutate(bb.sp = factor(bb.sp),
         yday = yday(date),
         elev = elev.mean/10) 
```

## 1.2 Visualize
```{r}
# Compare mean bias by criterion
ggplot(bias, aes(criterion, dprime)) +
  geom_boxplot()

# Bias ~ elevation * criterion
ggplot(bias, aes(elev.mean, dprime, color = criterion)) +
  geom_point(alpha = 0.25) +
  geom_smooth(se = TRUE)

ggplot(bias, aes(yday, dprime, color = criterion)) +
  geom_point(alpha = 0.25) +
  geom_smooth(se = TRUE)

ggplot(bias, aes(elev.mean, dprime, color = criterion)) +
  geom_point(alpha = 0.25) +
  geom_smooth(se = FALSE, method = "gam", formula = y ~ s(x, k = 5)) +
  facet_wrap(~bb.sp)

ggplot(bias, aes(yday, dprime, color = criterion)) +
  geom_point(alpha = 0.25) +
  geom_smooth(se = FALSE, method = "gam", formula = y ~ s(x, k = 5)) +
  facet_wrap(~bb.sp)

# Bias ~ year; Blüthgen's d' seems very robust to sampling effects, since it does not vary across years
ggplot(bias, aes(year, dprime, color = criterion)) +
  geom_boxplot() 

ggplot(bias, aes(year, dprime, color = criterion)) +
  geom_boxplot() +
  facet_wrap(~bb.sp)

# Does bias vary with abundance?
ggplot(bias, aes(log(bb.sp.abund), dprime, color = criterion)) +
  #geom_point() +
  geom_smooth() +
  facet_wrap(~bb.sp)

ggplot(bias, aes(bb.sp.abund, dprime, color = criterion)) +
  #geom_point() +
  geom_smooth() +
  facet_wrap(~bb.sp, )

ggplot(bias, aes(log(bb.sp.abund), dprime, color = criterion)) +
  geom_point(alpha = 0.1) +
  geom_smooth()

ggplot(bias_plantsp.sitedate, aes(log(bb.sp.prop.abund), dprime)) +
  geom_point() +
  geom_smooth() +
  facet_wrap(~bb.sp)
```

## 1.3 Multi-level model: bias ~ elevation + (1 + elevation | bb.sp)
```{r}
# brms
brm.ktype1 <- brm(dprime ~ yday + elev + (1 + elev | bb.sp), 
                  control = list(max_treedepth = 10),
                  cores = 3,
                  iter = 2000,
                  file = "./output/brm_ktype1",
                  data = filter(bias, criterion ==  "ktype"))

summary(brm.ktype1)
plot(brm.ktype1)
parnames(brm.ktype1)
mcmc_plot(brm.ktype1, pars = c("^r_"))
mcmc_plot(brm.ktype1, pars = c("^b_"))
mcmc_plot(brm.ktype1)
conditional_effects(brm.ktype1)
pp_check(brm.ktype1)

brm.ktype2 <- brm(dprime ~ 1 + yday + elev + (1 | bb.sp), 
                  control = list(max_treedepth = 10),
                  cores = 4,
                  iter = 2000,
                  file = "./output/brm_ktype2b",
                  data = filter(bias, criterion ==  "ktype"))

summary(brm.ktype2)
plot(brm.ktype2)
parnames(brm.ktype2)
mcmc_plot(brm.ktype2, pars = c("^r_"))
mcmc_plot(brm.ktype2, pars = c("^b_"))
mcmc_plot(brm.ktype2)
conditional_effects(brm.ktype2)
pp_check(brm.ktype2)

brm.ktype3 <- brm(dprime ~ 1 + (1 | bb.sp), 
                  control = list(max_treedepth = 10),
                  cores = 4,
                  iter = 2000,
                  family = "poisson",
                  file = "./output/brm_ktype3",
                  data = filter(bias, criterion ==  "ktype"))

summary(brm.ktype3)
plot(brm.ktype3)
parnames(brm.ktype3)
mcmc_plot(brm.ktype3, pars = c("^r_"))
mcmc_plot(brm.ktype3, pars = c("^b_"))
mcmc_plot(brm.ktype3)
conditional_effects(brm.ktype3)
pp_check(brm.ktype3)


# lme4
library(lme4)
library(ggeffects)

lmer.ktype <- lmer(dprime ~ 1 + elev:yday + (1 | bb.sp), 
                  data = filter(bias, criterion ==  "ktype"))
summary(lmer.ktype)
pred.ktype_yday <- ggpredict(lmer.ktype, "yday")
pred.ktype_elev <- ggpredict(lmer.ktype, "elev")
#pred.ktype_yday_ran <- ggpredict(lmer.ktype, "yday", type = "random")
#pred.ktype_elev_ran <- ggpredict(lmer.ktype, "elev", type = "random")
pred.ktype_bbsp <- ggpredict(lmer.ktype, "bb.sp", type = "random")

plot(pred.ktype_yday)
plot(pred.ktype_elev)
#plot(pred.ktype_yday_ran)
#plot(pred.ktype_elev_ran)
plot(pred.ktype_bbsp)



# mgcv
gam.ktype <- bam(dprime ~ s(bb.sp, bs = "re") + 
                   t2(yday, elev.mean, bs = c("tp", "gp")),
                 method = "fREML",
                 discrete = TRUE,
                 data = filter(bias, criterion ==  "ktype")) %>% getViz()

summary(gam.ktype)
plot(gam.ktype, allTerms = TRUE)

# inla
library(INLA)

inla.ktype <- inla(dprime ~ elev.mean, 
                   family = "gaussian",
                   data = filter(bias, criterion ==  "ktype"))

inla.pen <- inla(dprime ~ 1 + elev.mean,
  data = filter(bias, criterion ==  "ktype"), control.predictor = list(compute = TRUE))
summary(inla.pen)
```

## 1.4 Define hypotheses
```{r}
# Is d' different from 0?
H1 <- paste("Intercept = 0")

# Are some species more biased in their foraging than others? 
H.bss_gers <- paste("r_bb.sp[bss,Intercept] = r_bb.sp[gers,Intercept]")
H.bss_hort <- paste("r_bb.sp[bss,Intercept] = r_bb.sp[hort,Intercept]")
H.bss_humi <- paste("r_bb.sp[bss,Intercept] = r_bb.sp[humi,Intercept]")
H.bss_hypn <- paste("r_bb.sp[bss,Intercept] = r_bb.sp[hypn,Intercept]")
H.bss_jone <- paste("r_bb.sp[bss,Intercept] = r_bb.sp[jone,Intercept]")
H.bss_lapi <- paste("r_bb.sp[bss,Intercept] = r_bb.sp[lapi,Intercept]")
H.bss_mend <- paste("r_bb.sp[bss,Intercept] = r_bb.sp[mend,Intercept]")
H.bss_mont <- paste("r_bb.sp[bss,Intercept] = r_bb.sp[mont,Intercept]")
H.bss_muci <- paste("r_bb.sp[bss,Intercept] = r_bb.sp[muci,Intercept]")
H.bss_pasc <- paste("r_bb.sp[bss,Intercept] = r_bb.sp[pasc,Intercept]")
H.bss_prat <- paste("r_bb.sp[bss,Intercept] = r_bb.sp[prat,Intercept]")
H.bss_psit <- paste("r_bb.sp[bss,Intercept] = r_bb.sp[psit,Intercept]")
H.bss_pyre <- paste("r_bb.sp[bss,Intercept] = r_bb.sp[pyre,Intercept]")
H.bss_soro <- paste("r_bb.sp[bss,Intercept] = r_bb.sp[soro,Intercept]")
H.bss_wurf <- paste("r_bb.sp[bss,Intercept] = r_bb.sp[wurf,Intercept]")

H.gers_hort <- paste("r_bb.sp[gers,Intercept] = r_bb.sp[hort,Intercept]")
H.gers_humi <- paste("r_bb.sp[gers,Intercept] = r_bb.sp[humi,Intercept]")
H.gers_hypn <- paste("r_bb.sp[gers,Intercept] = r_bb.sp[hypn,Intercept]")
H.gers_jone <- paste("r_bb.sp[gers,Intercept] = r_bb.sp[jone,Intercept]")
H.gers_lapi <- paste("r_bb.sp[gers,Intercept] = r_bb.sp[lapi,Intercept]")
H.gers_mend <- paste("r_bb.sp[gers,Intercept] = r_bb.sp[mend,Intercept]")
H.gers_mont <- paste("r_bb.sp[gers,Intercept] = r_bb.sp[mont,Intercept]")
H.gers_muci <- paste("r_bb.sp[gers,Intercept] = r_bb.sp[muci,Intercept]")
H.gers_pasc <- paste("r_bb.sp[gers,Intercept] = r_bb.sp[pasc,Intercept]")
H.gers_prat <- paste("r_bb.sp[gers,Intercept] = r_bb.sp[prat,Intercept]")
H.gers_psit <- paste("r_bb.sp[gers,Intercept] = r_bb.sp[psit,Intercept]")
H.gers_pyre <- paste("r_bb.sp[gers,Intercept] = r_bb.sp[pyre,Intercept]")
H.gers_soro <- paste("r_bb.sp[gers,Intercept] = r_bb.sp[soro,Intercept]")
H.gers_wurf <- paste("r_bb.sp[gers,Intercept] = r_bb.sp[wurf,Intercept]")

H.hort_humi <- paste("r_bb.sp[hort,Intercept] = r_bb.sp[humi,Intercept]")
H.hort_hypn <- paste("r_bb.sp[hort,Intercept] = r_bb.sp[hypn,Intercept]")
H.hort_jone <- paste("r_bb.sp[hort,Intercept] = r_bb.sp[jone,Intercept]")
H.hort_lapi <- paste("r_bb.sp[hort,Intercept] = r_bb.sp[lapi,Intercept]")
H.hort_mend <- paste("r_bb.sp[hort,Intercept] = r_bb.sp[mend,Intercept]")
H.hort_mont <- paste("r_bb.sp[hort,Intercept] = r_bb.sp[mont,Intercept]")
H.hort_muci <- paste("r_bb.sp[hort,Intercept] = r_bb.sp[muci,Intercept]")
H.hort_pasc <- paste("r_bb.sp[hort,Intercept] = r_bb.sp[pasc,Intercept]")
H.hort_prat <- paste("r_bb.sp[hort,Intercept] = r_bb.sp[prat,Intercept]")
H.hort_psit <- paste("r_bb.sp[hort,Intercept] = r_bb.sp[psit,Intercept]")
H.hort_pyre <- paste("r_bb.sp[hort,Intercept] = r_bb.sp[pyre,Intercept]")
H.hort_soro <- paste("r_bb.sp[hort,Intercept] = r_bb.sp[soro,Intercept]")
H.hort_wurf <- paste("r_bb.sp[hort,Intercept] = r_bb.sp[wurf,Intercept]")

H.humi_hypn <- paste("r_bb.sp[humi,Intercept] = r_bb.sp[hypn,Intercept]")
H.humi_jone <- paste("r_bb.sp[humi,Intercept] = r_bb.sp[jone,Intercept]")
H.humi_lapi <- paste("r_bb.sp[humi,Intercept] = r_bb.sp[lapi,Intercept]")
H.humi_mend <- paste("r_bb.sp[humi,Intercept] = r_bb.sp[mend,Intercept]")
H.humi_mont <- paste("r_bb.sp[humi,Intercept] = r_bb.sp[mont,Intercept]")
H.humi_muci <- paste("r_bb.sp[humi,Intercept] = r_bb.sp[muci,Intercept]")
H.humi_pasc <- paste("r_bb.sp[humi,Intercept] = r_bb.sp[pasc,Intercept]")
H.humi_prat <- paste("r_bb.sp[humi,Intercept] = r_bb.sp[prat,Intercept]")
H.humi_psit <- paste("r_bb.sp[humi,Intercept] = r_bb.sp[psit,Intercept]")
H.humi_pyre <- paste("r_bb.sp[humi,Intercept] = r_bb.sp[pyre,Intercept]")
H.humi_soro <- paste("r_bb.sp[humi,Intercept] = r_bb.sp[soro,Intercept]")
H.humi_wurf <- paste("r_bb.sp[humi,Intercept] = r_bb.sp[wurf,Intercept]")

H.hypn_jone <- paste("r_bb.sp[hypn,Intercept] = r_bb.sp[jone,Intercept]")
H.hypn_lapi <- paste("r_bb.sp[hypn,Intercept] = r_bb.sp[lapi,Intercept]")
H.hypn_mend <- paste("r_bb.sp[hypn,Intercept] = r_bb.sp[mend,Intercept]")
H.hypn_mont <- paste("r_bb.sp[hypn,Intercept] = r_bb.sp[mont,Intercept]")
H.hypn_muci <- paste("r_bb.sp[hypn,Intercept] = r_bb.sp[muci,Intercept]")
H.hypn_pasc <- paste("r_bb.sp[hypn,Intercept] = r_bb.sp[pasc,Intercept]")
H.hypn_prat <- paste("r_bb.sp[hypn,Intercept] = r_bb.sp[prat,Intercept]")
H.hypn_psit <- paste("r_bb.sp[hypn,Intercept] = r_bb.sp[psit,Intercept]")
H.hypn_pyre <- paste("r_bb.sp[hypn,Intercept] = r_bb.sp[pyre,Intercept]")
H.hypn_soro <- paste("r_bb.sp[hypn,Intercept] = r_bb.sp[soro,Intercept]")
H.hypn_wurf <- paste("r_bb.sp[hypn,Intercept] = r_bb.sp[wurf,Intercept]")

H.jone_lapi <- paste("r_bb.sp[jone,Intercept] = r_bb.sp[lapi,Intercept]")
H.jone_mend <- paste("r_bb.sp[jone,Intercept] = r_bb.sp[mend,Intercept]")
H.jone_mont <- paste("r_bb.sp[jone,Intercept] = r_bb.sp[mont,Intercept]")
H.jone_muci <- paste("r_bb.sp[jone,Intercept] = r_bb.sp[muci,Intercept]")
H.jone_pasc <- paste("r_bb.sp[jone,Intercept] = r_bb.sp[pasc,Intercept]")
H.jone_prat <- paste("r_bb.sp[jone,Intercept] = r_bb.sp[prat,Intercept]")
H.jone_psit <- paste("r_bb.sp[jone,Intercept] = r_bb.sp[psit,Intercept]")
H.jone_pyre <- paste("r_bb.sp[jone,Intercept] = r_bb.sp[pyre,Intercept]")
H.jone_soro <- paste("r_bb.sp[jone,Intercept] = r_bb.sp[soro,Intercept]")
H.jone_wurf <- paste("r_bb.sp[jone,Intercept] = r_bb.sp[wurf,Intercept]")

H.lapi_mend <- paste("r_bb.sp[lapi,Intercept] = r_bb.sp[mend,Intercept]")
H.lapi_mont <- paste("r_bb.sp[lapi,Intercept] = r_bb.sp[mont,Intercept]")
H.lapi_muci <- paste("r_bb.sp[lapi,Intercept] = r_bb.sp[muci,Intercept]")
H.lapi_pasc <- paste("r_bb.sp[lapi,Intercept] = r_bb.sp[pasc,Intercept]")
H.lapi_prat <- paste("r_bb.sp[lapi,Intercept] = r_bb.sp[prat,Intercept]")
H.lapi_psit <- paste("r_bb.sp[lapi,Intercept] = r_bb.sp[psit,Intercept]")
H.lapi_pyre <- paste("r_bb.sp[lapi,Intercept] = r_bb.sp[pyre,Intercept]")
H.lapi_soro <- paste("r_bb.sp[lapi,Intercept] = r_bb.sp[soro,Intercept]")
H.lapi_wurf <- paste("r_bb.sp[lapi,Intercept] = r_bb.sp[wurf,Intercept]")

H.mend_mont <- paste("r_bb.sp[mend,Intercept] = r_bb.sp[mont,Intercept]")
H.mend_muci <- paste("r_bb.sp[mend,Intercept] = r_bb.sp[muci,Intercept]")
H.mend_pasc <- paste("r_bb.sp[mend,Intercept] = r_bb.sp[pasc,Intercept]")
H.mend_prat <- paste("r_bb.sp[mend,Intercept] = r_bb.sp[prat,Intercept]")
H.mend_psit <- paste("r_bb.sp[mend,Intercept] = r_bb.sp[psit,Intercept]")
H.mend_pyre <- paste("r_bb.sp[mend,Intercept] = r_bb.sp[pyre,Intercept]")
H.mend_soro <- paste("r_bb.sp[mend,Intercept] = r_bb.sp[soro,Intercept]")
H.mend_wurf <- paste("r_bb.sp[mend,Intercept] = r_bb.sp[wurf,Intercept]")

H.mont_muci <- paste("r_bb.sp[mont,Intercept] = r_bb.sp[muci,Intercept]")
H.mont_pasc <- paste("r_bb.sp[mont,Intercept] = r_bb.sp[pasc,Intercept]")
H.mont_prat <- paste("r_bb.sp[mont,Intercept] = r_bb.sp[prat,Intercept]")
H.mont_psit <- paste("r_bb.sp[mont,Intercept] = r_bb.sp[psit,Intercept]")
H.mont_pyre <- paste("r_bb.sp[mont,Intercept] = r_bb.sp[pyre,Intercept]")
H.mont_soro <- paste("r_bb.sp[mont,Intercept] = r_bb.sp[soro,Intercept]")
H.mont_wurf <- paste("r_bb.sp[mont,Intercept] = r_bb.sp[wurf,Intercept]")

H.muci_pasc <- paste("r_bb.sp[muci,Intercept] = r_bb.sp[pasc,Intercept]")
H.muci_prat <- paste("r_bb.sp[muci,Intercept] = r_bb.sp[prat,Intercept]")
H.muci_psit <- paste("r_bb.sp[muci,Intercept] = r_bb.sp[psit,Intercept]")
H.muci_pyre <- paste("r_bb.sp[muci,Intercept] = r_bb.sp[pyre,Intercept]")
H.muci_soro <- paste("r_bb.sp[muci,Intercept] = r_bb.sp[soro,Intercept]")
H.muci_wurf <- paste("r_bb.sp[muci,Intercept] = r_bb.sp[wurf,Intercept]")


H.pasc_prat <- paste("r_bb.sp[pasc,Intercept] = r_bb.sp[prat,Intercept]")
H.pasc_psit <- paste("r_bb.sp[pasc,Intercept] = r_bb.sp[psit,Intercept]")
H.pasc_pyre <- paste("r_bb.sp[pasc,Intercept] = r_bb.sp[pyre,Intercept]")
H.pasc_soro <- paste("r_bb.sp[pasc,Intercept] = r_bb.sp[soro,Intercept]")
H.pasc_wurf <- paste("r_bb.sp[pasc,Intercept] = r_bb.sp[wurf,Intercept]")


H.prat_psit <- paste("r_bb.sp[prat,Intercept] = r_bb.sp[psit,Intercept]")
H.prat_pyre <- paste("r_bb.sp[prat,Intercept] = r_bb.sp[pyre,Intercept]")
H.prat_soro <- paste("r_bb.sp[prat,Intercept] = r_bb.sp[soro,Intercept]")
H.prat_wurf <- paste("r_bb.sp[prat,Intercept] = r_bb.sp[wurf,Intercept]")


H.psit_pyre <- paste("r_bb.sp[psit,Intercept] = r_bb.sp[pyre,Intercept]")
H.psit_soro <- paste("r_bb.sp[psit,Intercept] = r_bb.sp[soro,Intercept]")
H.psit_wurf <- paste("r_bb.sp[psit,Intercept] = r_bb.sp[wurf,Intercept]")

H.pyre_soro <- paste("r_bb.sp[pyre,Intercept] = r_bb.sp[soro,Intercept]")
H.pyre_wurf <- paste("r_bb.sp[pyre,Intercept] = r_bb.sp[wurf,Intercept]")

H.soro_wurf <- paste("r_bb.sp[soro,Intercept] = r_bb.sp[wurf,Intercept]")
```

## 1.5 Test contrasts
```{r}
### Plant.sp
hypothesis(brm.plantsp, H1)

hypothesis(brm.plantsp, H.bss_gers, class = "")
hypothesis(brm.plantsp, H.bss_hort, class = "")
hypothesis(brm.plantsp, H.bss_humi, class = "")
hypothesis(brm.plantsp, H.bss_hypn, class = "")
hypothesis(brm.plantsp, H.bss_jone, class = "")
hypothesis(brm.plantsp, H.bss_lapi, class = "")
hypothesis(brm.plantsp, H.bss_mend, class = "")
hypothesis(brm.plantsp, H.bss_mont, class = "")
hypothesis(brm.plantsp, H.bss_muci, class = "")
hypothesis(brm.plantsp, H.bss_pasc, class = "")
hypothesis(brm.plantsp, H.bss_prat, class = "")
hypothesis(brm.plantsp, H.bss_psit, class = "")
hypothesis(brm.plantsp, H.bss_pyre, class = "")
hypothesis(brm.plantsp, H.bss_soro, class = "")
hypothesis(brm.plantsp, H.bss_wurf, class = "")

hypothesis(brm.plantsp, H.gers_hort, class = "")
hypothesis(brm.plantsp, H.gers_humi, class = "")
hypothesis(brm.plantsp, H.gers_hypn, class = "")
hypothesis(brm.plantsp, H.gers_jone, class = "")
hypothesis(brm.plantsp, H.gers_lapi, class = "")
hypothesis(brm.plantsp, H.gers_mend, class = "")
hypothesis(brm.plantsp, H.gers_mont, class = "")
hypothesis(brm.plantsp, H.gers_muci, class = "")
hypothesis(brm.plantsp, H.gers_pasc, class = "")
hypothesis(brm.plantsp, H.gers_prat, class = "")
hypothesis(brm.plantsp, H.gers_psit, class = "")
hypothesis(brm.plantsp, H.gers_pyre, class = "")
hypothesis(brm.plantsp, H.gers_soro, class = "")
hypothesis(brm.plantsp, H.gers_wurf, class = "")

hypothesis(brm.plantsp, H.hort_humi, class = "")
hypothesis(brm.plantsp, H.hort_hypn, class = "")
hypothesis(brm.plantsp, H.hort_jone, class = "")
hypothesis(brm.plantsp, H.hort_lapi, class = "")
hypothesis(brm.plantsp, H.hort_mend, class = "")
hypothesis(brm.plantsp, H.hort_mont, class = "")
hypothesis(brm.plantsp, H.hort_muci, class = "")
hypothesis(brm.plantsp, H.hort_pasc, class = "")
hypothesis(brm.plantsp, H.hort_prat, class = "")
hypothesis(brm.plantsp, H.hort_psit, class = "")
hypothesis(brm.plantsp, H.hort_pyre, class = "")
hypothesis(brm.plantsp, H.hort_soro, class = "")
hypothesis(brm.plantsp, H.hort_wurf, class = "")

hypothesis(brm.plantsp, H.humi_hypn, class = "")
hypothesis(brm.plantsp, H.humi_jone, class = "")
hypothesis(brm.plantsp, H.humi_lapi, class = "")
hypothesis(brm.plantsp, H.humi_mend, class = "")
hypothesis(brm.plantsp, H.humi_mont, class = "")
hypothesis(brm.plantsp, H.humi_muci, class = "")
hypothesis(brm.plantsp, H.humi_pasc, class = "")
hypothesis(brm.plantsp, H.humi_prat, class = "")
hypothesis(brm.plantsp, H.humi_psit, class = "")
hypothesis(brm.plantsp, H.humi_pyre, class = "")
hypothesis(brm.plantsp, H.humi_soro, class = "")
hypothesis(brm.plantsp, H.humi_wurf, class = "")

hypothesis(brm.plantsp, H.hypn_jone, class = "")
hypothesis(brm.plantsp, H.hypn_lapi, class = "")
hypothesis(brm.plantsp, H.hypn_mend, class = "")
hypothesis(brm.plantsp, H.hypn_mont, class = "")
hypothesis(brm.plantsp, H.hypn_muci, class = "")
hypothesis(brm.plantsp, H.hypn_pasc, class = "")
hypothesis(brm.plantsp, H.hypn_prat, class = "")
hypothesis(brm.plantsp, H.hypn_psit, class = "")
hypothesis(brm.plantsp, H.hypn_pyre, class = "")
hypothesis(brm.plantsp, H.hypn_soro, class = "")
hypothesis(brm.plantsp, H.hypn_wurf, class = "")

hypothesis(brm.plantsp, H.jone_lapi, class = "")
hypothesis(brm.plantsp, H.jone_mend, class = "")
hypothesis(brm.plantsp, H.jone_mont, class = "")
hypothesis(brm.plantsp, H.jone_muci, class = "")
hypothesis(brm.plantsp, H.jone_pasc, class = "")
hypothesis(brm.plantsp, H.jone_prat, class = "")
hypothesis(brm.plantsp, H.jone_psit, class = "")
hypothesis(brm.plantsp, H.jone_pyre, class = "")
hypothesis(brm.plantsp, H.jone_soro, class = "")
hypothesis(brm.plantsp, H.jone_wurf, class = "")

hypothesis(brm.plantsp, H.lapi_mend, class = "")
hypothesis(brm.plantsp, H.lapi_mont, class = "")
hypothesis(brm.plantsp, H.lapi_muci, class = "")
hypothesis(brm.plantsp, H.lapi_pasc, class = "")
hypothesis(brm.plantsp, H.lapi_prat, class = "")
hypothesis(brm.plantsp, H.lapi_psit, class = "")
hypothesis(brm.plantsp, H.lapi_pyre, class = "")
hypothesis(brm.plantsp, H.lapi_soro, class = "")
hypothesis(brm.plantsp, H.lapi_wurf, class = "")

hypothesis(brm.plantsp, H.mend_mont, class = "")
hypothesis(brm.plantsp, H.mend_muci, class = "")
hypothesis(brm.plantsp, H.mend_pasc, class = "")
hypothesis(brm.plantsp, H.mend_prat, class = "")
hypothesis(brm.plantsp, H.mend_psit, class = "")
hypothesis(brm.plantsp, H.mend_pyre, class = "")
hypothesis(brm.plantsp, H.mend_soro, class = "")
hypothesis(brm.plantsp, H.mend_wurf, class = "")

hypothesis(brm.plantsp, H.mont_muci, class = "")
hypothesis(brm.plantsp, H.mont_pasc, class = "")
hypothesis(brm.plantsp, H.mont_prat, class = "")
hypothesis(brm.plantsp, H.mont_psit, class = "")
hypothesis(brm.plantsp, H.mont_pyre, class = "")
hypothesis(brm.plantsp, H.mont_soro, class = "")
hypothesis(brm.plantsp, H.mont_wurf, class = "")

hypothesis(brm.plantsp, H.muci_pasc, class = "")
hypothesis(brm.plantsp, H.muci_prat, class = "")
hypothesis(brm.plantsp, H.muci_psit, class = "")
hypothesis(brm.plantsp, H.muci_pyre, class = "")
hypothesis(brm.plantsp, H.muci_soro, class = "")
hypothesis(brm.plantsp, H.muci_wurf, class = "")

hypothesis(brm.plantsp, H.pasc_prat, class = "")
hypothesis(brm.plantsp, H.pasc_psit, class = "")
hypothesis(brm.plantsp, H.pasc_pyre, class = "")
hypothesis(brm.plantsp, H.pasc_soro, class = "")
hypothesis(brm.plantsp, H.pasc_wurf, class = "")

hypothesis(brm.plantsp, H.prat_psit, class = "")
hypothesis(brm.plantsp, H.prat_pyre, class = "")
hypothesis(brm.plantsp, H.prat_soro, class = "")
hypothesis(brm.plantsp, H.prat_wurf, class = "")

hypothesis(brm.plantsp, H.psit_pyre, class = "")
hypothesis(brm.plantsp, H.psit_soro, class = "")
hypothesis(brm.plantsp, H.psit_wurf, class = "")

hypothesis(brm.plantsp, H.pyre_soro, class = "")
hypothesis(brm.plantsp, H.pyre_wurf, class = "")

hypothesis(brm.plantsp, H.soro_wurf, class = "")

### Kugler type
hypothesis(brm.ktype, H1)

hypothesis(brm.ktype, H.bss_gers, class = "")
hypothesis(brm.ktype, H.bss_hort, class = "")
hypothesis(brm.ktype, H.bss_humi, class = "")
hypothesis(brm.ktype, H.bss_hypn, class = "")
hypothesis(brm.ktype, H.bss_jone, class = "")
hypothesis(brm.ktype, H.bss_lapi, class = "")
hypothesis(brm.ktype, H.bss_mend, class = "")
hypothesis(brm.ktype, H.bss_mont, class = "")
hypothesis(brm.ktype, H.bss_muci, class = "")
hypothesis(brm.ktype, H.bss_pasc, class = "")
hypothesis(brm.ktype, H.bss_prat, class = "")
hypothesis(brm.ktype, H.bss_psit, class = "")
hypothesis(brm.ktype, H.bss_pyre, class = "")
hypothesis(brm.ktype, H.bss_soro, class = "")
hypothesis(brm.ktype, H.bss_wurf, class = "")

hypothesis(brm.ktype, H.gers_hort, class = "")
hypothesis(brm.ktype, H.gers_humi, class = "")
hypothesis(brm.ktype, H.gers_hypn, class = "")
hypothesis(brm.ktype, H.gers_jone, class = "")
hypothesis(brm.ktype, H.gers_lapi, class = "")
hypothesis(brm.ktype, H.gers_mend, class = "")
hypothesis(brm.ktype, H.gers_mont, class = "")
hypothesis(brm.ktype, H.gers_muci, class = "")
hypothesis(brm.ktype, H.gers_pasc, class = "")
hypothesis(brm.ktype, H.gers_prat, class = "")
hypothesis(brm.ktype, H.gers_psit, class = "")
hypothesis(brm.ktype, H.gers_pyre, class = "")
hypothesis(brm.ktype, H.gers_soro, class = "")
hypothesis(brm.ktype, H.gers_wurf, class = "")

hypothesis(brm.ktype, H.hort_humi, class = "")
hypothesis(brm.ktype, H.hort_hypn, class = "")
hypothesis(brm.ktype, H.hort_jone, class = "")
hypothesis(brm.ktype, H.hort_lapi, class = "")
hypothesis(brm.ktype, H.hort_mend, class = "")
hypothesis(brm.ktype, H.hort_mont, class = "")
hypothesis(brm.ktype, H.hort_muci, class = "")
hypothesis(brm.ktype, H.hort_pasc, class = "")
hypothesis(brm.ktype, H.hort_prat, class = "")
hypothesis(brm.ktype, H.hort_psit, class = "")
hypothesis(brm.ktype, H.hort_pyre, class = "")
hypothesis(brm.ktype, H.hort_soro, class = "")
hypothesis(brm.ktype, H.hort_wurf, class = "")

hypothesis(brm.ktype, H.humi_hypn, class = "")
hypothesis(brm.ktype, H.humi_jone, class = "")
hypothesis(brm.ktype, H.humi_lapi, class = "")
hypothesis(brm.ktype, H.humi_mend, class = "")
hypothesis(brm.ktype, H.humi_mont, class = "")
hypothesis(brm.ktype, H.humi_muci, class = "")
hypothesis(brm.ktype, H.humi_pasc, class = "")
hypothesis(brm.ktype, H.humi_prat, class = "")
hypothesis(brm.ktype, H.humi_psit, class = "")
hypothesis(brm.ktype, H.humi_pyre, class = "")
hypothesis(brm.ktype, H.humi_soro, class = "")
hypothesis(brm.ktype, H.humi_wurf, class = "")

hypothesis(brm.ktype, H.hypn_jone, class = "")
hypothesis(brm.ktype, H.hypn_lapi, class = "")
hypothesis(brm.ktype, H.hypn_mend, class = "")
hypothesis(brm.ktype, H.hypn_mont, class = "")
hypothesis(brm.ktype, H.hypn_muci, class = "")
hypothesis(brm.ktype, H.hypn_pasc, class = "")
hypothesis(brm.ktype, H.hypn_prat, class = "")
hypothesis(brm.ktype, H.hypn_psit, class = "")
hypothesis(brm.ktype, H.hypn_pyre, class = "")
hypothesis(brm.ktype, H.hypn_soro, class = "")
hypothesis(brm.ktype, H.hypn_wurf, class = "")

hypothesis(brm.ktype, H.jone_lapi, class = "")
hypothesis(brm.ktype, H.jone_mend, class = "")
hypothesis(brm.ktype, H.jone_mont, class = "")
hypothesis(brm.ktype, H.jone_muci, class = "")
hypothesis(brm.ktype, H.jone_pasc, class = "")
hypothesis(brm.ktype, H.jone_prat, class = "")
hypothesis(brm.ktype, H.jone_psit, class = "")
hypothesis(brm.ktype, H.jone_pyre, class = "")
hypothesis(brm.ktype, H.jone_soro, class = "")
hypothesis(brm.ktype, H.jone_wurf, class = "")

hypothesis(brm.ktype, H.lapi_mend, class = "")
hypothesis(brm.ktype, H.lapi_mont, class = "")
hypothesis(brm.ktype, H.lapi_muci, class = "")
hypothesis(brm.ktype, H.lapi_pasc, class = "")
hypothesis(brm.ktype, H.lapi_prat, class = "")
hypothesis(brm.ktype, H.lapi_psit, class = "")
hypothesis(brm.ktype, H.lapi_pyre, class = "")
hypothesis(brm.ktype, H.lapi_soro, class = "")
hypothesis(brm.ktype, H.lapi_wurf, class = "")

hypothesis(brm.ktype, H.mend_mont, class = "")
hypothesis(brm.ktype, H.mend_muci, class = "")
hypothesis(brm.ktype, H.mend_pasc, class = "")
hypothesis(brm.ktype, H.mend_prat, class = "")
hypothesis(brm.ktype, H.mend_psit, class = "")
hypothesis(brm.ktype, H.mend_pyre, class = "")
hypothesis(brm.ktype, H.mend_soro, class = "")
hypothesis(brm.ktype, H.mend_wurf, class = "")

hypothesis(brm.ktype, H.mont_muci, class = "")
hypothesis(brm.ktype, H.mont_pasc, class = "")
hypothesis(brm.ktype, H.mont_prat, class = "")
hypothesis(brm.ktype, H.mont_psit, class = "")
hypothesis(brm.ktype, H.mont_pyre, class = "")
hypothesis(brm.ktype, H.mont_soro, class = "")
hypothesis(brm.ktype, H.mont_wurf, class = "")

hypothesis(brm.ktype, H.muci_pasc, class = "")
hypothesis(brm.ktype, H.muci_prat, class = "")
hypothesis(brm.ktype, H.muci_psit, class = "")
hypothesis(brm.ktype, H.muci_pyre, class = "")
hypothesis(brm.ktype, H.muci_soro, class = "")
hypothesis(brm.ktype, H.muci_wurf, class = "")

hypothesis(brm.ktype, H.pasc_prat, class = "")
hypothesis(brm.ktype, H.pasc_psit, class = "")
hypothesis(brm.ktype, H.pasc_pyre, class = "")
hypothesis(brm.ktype, H.pasc_soro, class = "")
hypothesis(brm.ktype, H.pasc_wurf, class = "")

hypothesis(brm.ktype, H.prat_psit, class = "")
hypothesis(brm.ktype, H.prat_pyre, class = "")
hypothesis(brm.ktype, H.prat_soro, class = "")
hypothesis(brm.ktype, H.prat_wurf, class = "")

hypothesis(brm.ktype, H.psit_pyre, class = "")
hypothesis(brm.ktype, H.psit_soro, class = "")
hypothesis(brm.ktype, H.psit_wurf, class = "")

hypothesis(brm.ktype, H.pyre_soro, class = "")
hypothesis(brm.ktype, H.pyre_wurf, class = "")

hypothesis(brm.ktype, H.soro_wurf, class = "")

### Plant family
hypothesis(brm.family, H1)

hypothesis(brm.family, H.bss_gers, class = "")
hypothesis(brm.family, H.bss_hort, class = "")
hypothesis(brm.family, H.bss_humi, class = "")
hypothesis(brm.family, H.bss_hypn, class = "")
hypothesis(brm.family, H.bss_jone, class = "")
hypothesis(brm.family, H.bss_lapi, class = "")
hypothesis(brm.family, H.bss_mend, class = "")
hypothesis(brm.family, H.bss_mont, class = "")
hypothesis(brm.family, H.bss_muci, class = "")
hypothesis(brm.family, H.bss_pasc, class = "")
hypothesis(brm.family, H.bss_prat, class = "")
hypothesis(brm.family, H.bss_psit, class = "")
hypothesis(brm.family, H.bss_pyre, class = "")
hypothesis(brm.family, H.bss_soro, class = "")
hypothesis(brm.family, H.bss_wurf, class = "")

hypothesis(brm.family, H.gers_hort, class = "")
hypothesis(brm.family, H.gers_humi, class = "")
hypothesis(brm.family, H.gers_hypn, class = "")
hypothesis(brm.family, H.gers_jone, class = "")
hypothesis(brm.family, H.gers_lapi, class = "")
hypothesis(brm.family, H.gers_mend, class = "")
hypothesis(brm.family, H.gers_mont, class = "")
hypothesis(brm.family, H.gers_muci, class = "")
hypothesis(brm.family, H.gers_pasc, class = "")
hypothesis(brm.family, H.gers_prat, class = "")
hypothesis(brm.family, H.gers_psit, class = "")
hypothesis(brm.family, H.gers_pyre, class = "")
hypothesis(brm.family, H.gers_soro, class = "")
hypothesis(brm.family, H.gers_wurf, class = "")

hypothesis(brm.family, H.hort_humi, class = "")
hypothesis(brm.family, H.hort_hypn, class = "")
hypothesis(brm.family, H.hort_jone, class = "")
hypothesis(brm.family, H.hort_lapi, class = "")
hypothesis(brm.family, H.hort_mend, class = "")
hypothesis(brm.family, H.hort_mont, class = "")
hypothesis(brm.family, H.hort_muci, class = "")
hypothesis(brm.family, H.hort_pasc, class = "")
hypothesis(brm.family, H.hort_prat, class = "")
hypothesis(brm.family, H.hort_psit, class = "")
hypothesis(brm.family, H.hort_pyre, class = "")
hypothesis(brm.family, H.hort_soro, class = "")
hypothesis(brm.family, H.hort_wurf, class = "")

hypothesis(brm.family, H.humi_hypn, class = "")
hypothesis(brm.family, H.humi_jone, class = "")
hypothesis(brm.family, H.humi_lapi, class = "")
hypothesis(brm.family, H.humi_mend, class = "")
hypothesis(brm.family, H.humi_mont, class = "")
hypothesis(brm.family, H.humi_muci, class = "")
hypothesis(brm.family, H.humi_pasc, class = "")
hypothesis(brm.family, H.humi_prat, class = "")
hypothesis(brm.family, H.humi_psit, class = "")
hypothesis(brm.family, H.humi_pyre, class = "")
hypothesis(brm.family, H.humi_soro, class = "")
hypothesis(brm.family, H.humi_wurf, class = "")

hypothesis(brm.family, H.hypn_jone, class = "")
hypothesis(brm.family, H.hypn_lapi, class = "")
hypothesis(brm.family, H.hypn_mend, class = "")
hypothesis(brm.family, H.hypn_mont, class = "")
hypothesis(brm.family, H.hypn_muci, class = "")
hypothesis(brm.family, H.hypn_pasc, class = "")
hypothesis(brm.family, H.hypn_prat, class = "")
hypothesis(brm.family, H.hypn_psit, class = "")
hypothesis(brm.family, H.hypn_pyre, class = "")
hypothesis(brm.family, H.hypn_soro, class = "")
hypothesis(brm.family, H.hypn_wurf, class = "")

hypothesis(brm.family, H.jone_lapi, class = "")
hypothesis(brm.family, H.jone_mend, class = "")
hypothesis(brm.family, H.jone_mont, class = "")
hypothesis(brm.family, H.jone_muci, class = "")
hypothesis(brm.family, H.jone_pasc, class = "")
hypothesis(brm.family, H.jone_prat, class = "")
hypothesis(brm.family, H.jone_psit, class = "")
hypothesis(brm.family, H.jone_pyre, class = "")
hypothesis(brm.family, H.jone_soro, class = "")
hypothesis(brm.family, H.jone_wurf, class = "")

hypothesis(brm.family, H.lapi_mend, class = "")
hypothesis(brm.family, H.lapi_mont, class = "")
hypothesis(brm.family, H.lapi_muci, class = "")
hypothesis(brm.family, H.lapi_pasc, class = "")
hypothesis(brm.family, H.lapi_prat, class = "")
hypothesis(brm.family, H.lapi_psit, class = "")
hypothesis(brm.family, H.lapi_pyre, class = "")
hypothesis(brm.family, H.lapi_soro, class = "")
hypothesis(brm.family, H.lapi_wurf, class = "")

hypothesis(brm.family, H.mend_mont, class = "")
hypothesis(brm.family, H.mend_muci, class = "")
hypothesis(brm.family, H.mend_pasc, class = "")
hypothesis(brm.family, H.mend_prat, class = "")
hypothesis(brm.family, H.mend_psit, class = "")
hypothesis(brm.family, H.mend_pyre, class = "")
hypothesis(brm.family, H.mend_soro, class = "")
hypothesis(brm.family, H.mend_wurf, class = "")

hypothesis(brm.family, H.mont_muci, class = "")
hypothesis(brm.family, H.mont_pasc, class = "")
hypothesis(brm.family, H.mont_prat, class = "")
hypothesis(brm.family, H.mont_psit, class = "")
hypothesis(brm.family, H.mont_pyre, class = "")
hypothesis(brm.family, H.mont_soro, class = "")
hypothesis(brm.family, H.mont_wurf, class = "")

hypothesis(brm.family, H.muci_pasc, class = "")
hypothesis(brm.family, H.muci_prat, class = "")
hypothesis(brm.family, H.muci_psit, class = "")
hypothesis(brm.family, H.muci_pyre, class = "")
hypothesis(brm.family, H.muci_soro, class = "")
hypothesis(brm.family, H.muci_wurf, class = "")

hypothesis(brm.family, H.pasc_prat, class = "")
hypothesis(brm.family, H.pasc_psit, class = "")
hypothesis(brm.family, H.pasc_pyre, class = "")
hypothesis(brm.family, H.pasc_soro, class = "")
hypothesis(brm.family, H.pasc_wurf, class = "")

hypothesis(brm.family, H.prat_psit, class = "")
hypothesis(brm.family, H.prat_pyre, class = "")
hypothesis(brm.family, H.prat_soro, class = "")
hypothesis(brm.family, H.prat_wurf, class = "")

hypothesis(brm.family, H.psit_pyre, class = "")
hypothesis(brm.family, H.psit_soro, class = "")
hypothesis(brm.family, H.psit_wurf, class = "")

hypothesis(brm.family, H.pyre_soro, class = "")
hypothesis(brm.family, H.pyre_wurf, class = "")

hypothesis(brm.family, H.soro_wurf, class = "")
```


# 2. Having established the bias of BB foraging, let's zoom in on which morphotypes and families were over/under-represented

## 2.1 Defining bias

Deciding exactly how to represent bias turns out to be rather tricky. I am trying out a few options right now.   

1. Bias could  be trated simply as the Euclidean difference between prop.flower.cover and prop.visits. This has the effect of weighting bias by flower.cover, particularly negative bias, since an abundant plant can have a greater negative bias (because it is farther from 0) than a rare plant. Euclidean bias is bounded between -1 and 1. Its symmetry is a nice property that sets it apart from proportional or fold-change bias. If, for whatever reason, I decided that a log-transformation of this measure were desirable, I think I could simply add a constant of 1 to all values, which would move the boundary up to (0, 2). I think the biggest problem with measuring bias in this way is that it makes no distinction between a plant that was visited a little and one that was visited not at all, and that could be an important biological difference. For example, consider two plants, one with a proportional abundance = 0.01 and proportional visitation = 0.01, and the second with proportional abundance = 0.01 and proportional visitation = 0.00. These will look identical in our analysis, but it is possible that the unvisited plant would not have been visited no matter how high its abundance was. **For starters, though, I am going to proceed with this metric because it is the simplest. And because DJ thought it was a good idea.**

2. Alternatively, bias could be calculated as a simple proportion: prop.visits/prop.flower.cover. If anything, this might tend to reverse the weighting, since a small Euclidean change in a rare species would be a large proportional change. Proportional change has a lower boundary of 0 and no upper boundary. This approach highlights cases of positive bias, but it is even worse than euclidean bias in terms of obscuring the distinction between plants visited slightly and those visited not at all. Worse in a sense. In another sense, it is better than euclidean bias, because with proportional visitation a plant visited 0 times is at least numerical distinguishable from one visited 1 time (visited plants will always have non-zero proportional bias). 

3. Similarly to proportional bias, I could calculate bias as fold change. This converges to be virtually identical to proportional change as bias becomes more positive, but fold change has a lower bound of -1 instead of 0. 

4. If the distinction between plants that were visited and those that were not is what we're most interested in, we could simply treat bias as binary. This, of course, comes at the price of not being able to distinguish a plant that received 0.01 proportional visitation and one that recevied  0.99 proportional visitation, and this distinction is arguably important. 

5. An interesting option is to first separate visited from unvisited plants, set the bias of unvisited plants to 0, and then exponentiate the euclidean bias of all visited plant species. This has some desirable properties. First, it clearly distinguished visited plants from unvisited ones. Second, it assigns plants visited perfectly in proportion to their abundance a sensible bias of 1 (just like proportional bias). Thus, plants with a psotive bias are > 1 (bounded at 2.718282 -- okay, so that's a bit inelegant), those with a negative bias are < 1 (bounded at 0.367895), and those not visited at all = 0. 

6. The problem with the option above is that is adds weight to positive bias. Another approach is to separate the unvisited plants from the visited, set unvisited bias = 0, and visited bias = prop.visits - prop.flower.cover + 1. The +1 serves to make the distribution positive, with 1 represening perfect proportionality between visitation and floral abundance. This creates a symmetrically bounded distribution between 0 (not inclusive) and 2 (inclusive), plus the true zeros from the unvisited plants. It feels a bit...slapped together, though.

## 2.2 Function for calculating bias per group (= criterion)
```{r}
biaseval_group <- function(x, criterion) {
  
  criterion <- enquo(criterion)
  
  x %>%
    mutate(criterion = factor(!!criterion)) %>% # having this as a factor will help with modeling later on
    group_by(site, date, bb.sp, criterion) %>%
    summarize(flower.cover = sum(flower.cover), # get raw floer cover and visits by group
              visits = sum(visits)) %>%
    group_by(site, date, bb.sp) %>%
    mutate(prop.flower.cover = flower.cover/sum(flower.cover), # get proportional flower cover and visits by group
           prop.visits = visits/sum(visits)) %>%
    mutate(bias = prop.visits - prop.flower.cover, # bias as difference
           bias.prop = prop.visits/prop.flower.cover, # bias as proportion
           bias.fold = bias/prop.flower.cover, # bias as fold change
           bias.bin = if_else(bias > 0, 1, 0), # bias as binary
           bias.exp = if_else(visits == 0, 0, exp(prop.visits - prop.flower.cover)),  # bias as exponentiated difference
           bias.split = if_else(visits == 0, 0, (prop.visits - prop.flower.cover + 1))) %>%
    mutate(yday = yday(date)) %>%
    select(site, date, yday, bb.sp, criterion, 
           bias, bias.prop, bias.fold, bias.bin, bias.exp, bias.split,
           visits, flower.cover, prop.visits, prop.flower.cover)
}
```

## 2.3 Calculate bias per plant speces/ktype/family
```{r}
bias_group_plantsp <- biaseval_group(visit_cover, plant.sp) %>%
  left_join(site_data) %>%
  rename(group = criterion) %>%
  mutate(criterion = rep("plant.sp", n()))

bias_group_ktype <- biaseval_group(visit_cover, k.type.ss) %>%
  left_join(site_data) %>%
  rename(group = criterion) %>%
  mutate(criterion = rep("ktype", n()))

bias_group_family <- biaseval_group(visit_cover, plant.family) %>%
  left_join(site_data) %>%
  rename(group = criterion) %>%
  mutate(criterion = rep("plant.family", n()))

bias_group <- bind_rows(bias_group_plantsp, bias_group_ktype, bias_group_family)

ggplot(bias_group, aes(prop.flower.cover, prop.visits)) +
  geom_point(alpha = 0.1) +
  #geom_smooth(method = "gam", formula = y ~ s(x, k = 4)) +
  geom_smooth(method = "lm") +
  geom_abline(slope = 1, color = "red", linetype = "dashed") +
  facet_wrap(~criterion)

ggplot(bias_group, aes(log(prop.flower.cover), prop.visits)) +
  geom_point(alpha = 0.1) +
  geom_smooth() +
  facet_wrap(~criterion)

ggplot(filter(bias_group, criterion == "ktype"), aes(prop.flower.cover, prop.visits)) +
  geom_point(alpha = 0.1) +
  geom_smooth() +
  geom_abline(slope = 1, color = "red", linetype = "dashed") +
  facet_wrap(~group)
```

## 2.4 Visualize
```{r}
# Check the distribution of bias
ggplot(bias_group_ktype, aes(bias)) +
  geom_histogram()

ggplot(bias_group_ktype, aes(bias)) +
  geom_histogram() +
  facet_wrap(~criterion)

ggplot(bias_group_family, aes(bias)) +
  geom_histogram()

ggplot(bias_group_family, aes(bias)) +
  geom_histogram() +
  facet_wrap(~criterion)

# Bias by Kugler morphotype
ggplot(bias_group_ktype, aes(group, bias)) +
  geom_boxplot()

ggplot(bias_group_ktype, aes(group, bias)) +
  geom_boxplot() +
  facet_wrap(~bb.sp)

ggplot(bias_group_ktype, aes(bb.sp, bias)) +
  geom_boxplot() +
  facet_wrap(~group)

ggplot(bias_group_ktype, aes(elev.mean, bias)) +
  #geom_point() +
  geom_smooth(method = gam, formula = y ~ s(x, k = 5)) +
  facet_wrap(~group)

# Bias by plant family
ggplot(bias_group_family, aes(group, bias)) +
  geom_boxplot()

ggplot(bias_group_family, aes(group, bias)) +
  geom_boxplot() +
  facet_wrap(~bb.sp)

ggplot(bias_group_family, aes(bb.sp, bias)) +
  geom_boxplot() +
  facet_wrap(~group)

ggplot(bias_group_family, aes(elev.mean, bias)) +
  #geom_point() +
  geom_smooth(method = gam, formula = y ~ s(x, k = 5)) +
  facet_wrap(~group)
```


## 2.5 Explaining bias by floral traits, BB spp, and elevation (and maybe time)
```{r}
repmod.ktype <- brm(bias ~ elev.mean + (1 | criterion), 
                    data = bias_ktype)

repmod.ktype <- brm(bias ~ criterion, 
                    data = bias_ktype)
parnames(repmod.ktype)
mcmc_plot(repmod.ktype)
conditional_effects(repmod.ktype)

H.K5_K6 <- paste("b_criterion5 = b_criterion6")
hypothesis(repmod.ktype, H.K5_K6, class = "")
```




# How do species vary in their biases
```{r}
bias_matrix.plantsp <- bias_group_plantsp %>%
  group_by(bb.sp, criterion) %>%
  summarize(bias = mean(bias)) %>%
  pivot_wider(values_from = bias, names_from = criterion, values_fill = 0) %>%
  column_to_rownames("bb.sp")

bias_matrix.ktype <- bias_group_ktype %>%
  group_by(bb.sp, criterion) %>%
  summarize(bias = mean(bias)) %>%
  pivot_wider(values_from = bias, names_from = criterion, values_fill = 0) %>%
  column_to_rownames("bb.sp")

library(vegan)

pca.plantsp <- rda(bias_matrix.plantsp)
plot(pca.plantsp, type = "t")

pca.ktype <- rda(bias_matrix.ktype)
plot(pca.ktype)


```


# 3. Can we explain bias by BB-Fl trait associations? (RLQ + fourth corner analysis)
## 3.1
## 3.2 Do BB-FL trait associations vary with elevation?

# 4. How does bias relate to network structure?
## 4.1 A function for calculating niche overlap
```{r}
# Niche overlap by sp.pair * site * year
niche_overlapper <- function(x, criterion) {
    
  criterion <- enquo(criterion) # criterion is set to either plant.sp (for taxonomic) or k.type.s (for morphological)
  
  no1 <- x %>%
    group_by(bb.sp, !!criterion, site, date) %>% # group by site-year and BB-FL pair
    summarize(visits = sum(visits)) %>%
    group_by(bb.sp, site, date) %>% # group by site-year and BB species
    mutate(prop.visits = visits/sum(visits)) %>% # get proportional interaction frequency for each BB-FL pair
    select(site, date, bb.sp, !!criterion, prop.visits) %>%
    pivot_wider(names_from = !!criterion, values_from = prop.visits, values_fill = 0) %>% # spread into the kind of array vegan likes
    group_by(site, date) %>%
    nest() %>% # nest into list-column
    mutate(data2 = map(data, function(x) column_to_rownames(x, var = "bb.sp")), # column to rownames
           dist = map(data2, function(x) vegdist(x, method = "horn")), # convert to Horn's distance matrix
           tidy_dist = map(dist, tidy)) %>% # convert to tidy array of Horn's distances
    dplyr::select(site, date, tidy_dist) %>% # drop intermediate list-columns
    unnest(cols = c(tidy_dist)) %>%
    rename(bb.sp1 = item1, bb.sp2 = item2) %>%
    mutate(niche.overlap = 1-distance,  # express niche overlap so that 1 = perfect overlap and 0 = no overlap
           bss = if_else(bb.sp1 == "bss" | bb.sp2 == "bss", TRUE, FALSE),
           pasc = if_else(bb.sp1 == "pasc" | bb.sp2 == "pasc", TRUE, FALSE),
           wurf = if_else(bb.sp1 == "wurf" | bb.sp2 == "wurf", TRUE, FALSE),
           prat = if_else(bb.sp1 == "prat" | bb.sp2 == "prat", TRUE, FALSE),
           soro = if_else(bb.sp1 == "soro" | bb.sp2 == "soro", TRUE, FALSE),
           humi = if_else(bb.sp1 == "humi" | bb.sp2 == "humi", TRUE, FALSE),
           muci = if_else(bb.sp1 == "muci" | bb.sp2 == "muci", TRUE, FALSE),
           pyre = if_else(bb.sp1 == "pyre" | bb.sp2 == "pyre", TRUE, FALSE),
           mont = if_else(bb.sp1 == "mont" | bb.sp2 == "mont", TRUE, FALSE),
           mend = if_else(bb.sp1 == "mend" | bb.sp2 == "mend", TRUE, FALSE),
           psit = if_else(bb.sp1 == "psit" | bb.sp2 == "psit", TRUE, FALSE),
           gers = if_else(bb.sp1 == "gers" | bb.sp2 == "gers", TRUE, FALSE),
           jone = if_else(bb.sp1 == "jone" | bb.sp2 == "jone", TRUE, FALSE),
           hypn = if_else(bb.sp1 == "hypn" | bb.sp2 == "hypn", TRUE, FALSE),
           lapi = if_else(bb.sp1 == "lapi" | bb.sp2 == "lapi", TRUE, FALSE),
           hort = if_else(bb.sp1 == "hort" | bb.sp2 == "hort", TRUE, FALSE))

  no <- list(
    
  no_bss <- no1 %>%
    group_by(site, date, bss) %>% 
    summarize(mean.niche.overlap = mean(niche.overlap),
              max.niche.overlap = max(niche.overlap)) %>%
    mutate(bb.sp = rep("bss", n())) %>%
    filter(bss == TRUE),
  
  no_pasc <- no1 %>%
    group_by(site, date, pasc) %>%
    summarize(mean.niche.overlap = mean(niche.overlap),
              max.niche.overlap = max(niche.overlap)) %>%    
    mutate(bb.sp = rep("pasc", n())) %>%
    filter(pasc == TRUE),
  
  no_wurf <- no1 %>%
    group_by(site, date, wurf) %>%
    summarize(mean.niche.overlap = mean(niche.overlap),
              max.niche.overlap = max(niche.overlap)) %>%    
    mutate(bb.sp = rep("wurf", n())) %>%
    filter(wurf == TRUE),
  
  no_prat <- no1 %>%
    group_by(site, date, prat) %>%
    summarize(mean.niche.overlap = mean(niche.overlap),
              max.niche.overlap = max(niche.overlap)) %>%   
    mutate(bb.sp = rep("prat", n())) %>%
    filter(prat == TRUE),
  
  no_soro <- no1 %>%
    group_by(site, date, soro) %>% 
    summarize(mean.niche.overlap = mean(niche.overlap),
              max.niche.overlap = max(niche.overlap)) %>%    
    mutate(bb.sp = rep("soro", n())) %>%
    filter(soro == TRUE),
  
  no_humi <- no1 %>%
    group_by(site, date, humi) %>%
    summarize(mean.niche.overlap = mean(niche.overlap),
              max.niche.overlap = max(niche.overlap)) %>%    
    mutate(bb.sp = rep("humi", n())) %>%
    filter(humi == TRUE),
  
  no_muci <- no1 %>%
    group_by(site, date, muci) %>%
    summarize(mean.niche.overlap = mean(niche.overlap),
              max.niche.overlap = max(niche.overlap)) %>%    
    mutate(bb.sp = rep("muci", n())) %>%
    filter(muci == TRUE),
  
  no_pyre <- no1 %>%
    group_by(site, date, pyre) %>%
    summarize(mean.niche.overlap = mean(niche.overlap),
              max.niche.overlap = max(niche.overlap)) %>%   
    mutate(bb.sp = rep("pyre", n())) %>%
    filter(pyre == TRUE),
  
  no_mont <- no1 %>%
    group_by(site, date, mont) %>% 
    summarize(mean.niche.overlap = mean(niche.overlap),
              max.niche.overlap = max(niche.overlap)) %>%    
    mutate(bb.sp = rep("mont", n())) %>%
    filter(mont == TRUE),
  
  no_mend <- no1 %>%
    group_by(site, date, mend) %>%
    summarize(mean.niche.overlap = mean(niche.overlap),
              max.niche.overlap = max(niche.overlap)) %>%    
    mutate(bb.sp = rep("mend", n())) %>%
    filter(mend == TRUE),
  
  no_psit <- no1 %>%
    group_by(site, date, psit) %>%
    summarize(mean.niche.overlap = mean(niche.overlap),
              max.niche.overlap = max(niche.overlap)) %>%    
    mutate(bb.sp = rep("psit", n())) %>%
    filter(psit == TRUE),
  
  no_gers <- no1 %>%
    group_by(site, date, gers) %>%
    summarize(mean.niche.overlap = mean(niche.overlap),
              max.niche.overlap = max(niche.overlap)) %>%    
    mutate(bb.sp = rep("gers", n())) %>%
    filter(gers == TRUE),
  
  no_jone <- no1 %>%
    group_by(site, date, jone) %>% 
    summarize(mean.niche.overlap = mean(niche.overlap),
              max.niche.overlap = max(niche.overlap)) %>%    
    mutate(bb.sp = rep("jone", n())) %>%
    filter(jone == TRUE),
  
  no_hypn <- no1 %>%
    group_by(site, date, hypn) %>%
    summarize(mean.niche.overlap = mean(niche.overlap),
              max.niche.overlap = max(niche.overlap)) %>%    
    mutate(bb.sp = rep("hypn", n())) %>%
    filter(hypn == TRUE),
  
  no_lapi <- no1 %>%
    group_by(site, date, lapi) %>%
    summarize(mean.niche.overlap = mean(niche.overlap),
              max.niche.overlap = max(niche.overlap)) %>%    
    mutate(bb.sp = rep("lapi", n())) %>%
    filter(lapi == TRUE),
  
  no_hort <- no1 %>%
    group_by(site, date, hort) %>%
    summarize(mean.niche.overlap = mean(niche.overlap),
              max.niche.overlap = max(niche.overlap)) %>%    
    mutate(bb.sp = rep("hort", n())) %>%
    filter(hort == TRUE)
  ) %>%
    
    bind_rows() %>%
    select(site, date, bb.sp, mean.niche.overlap, max.niche.overlap)
  }
```

## 4.2 Calculate niche overlap: NOT READY FOR INTERPRETATION YET
```{r}
niche_overlap_plant.sp <- niche_overlapper(visit_cover, plant.sp)

niche_overlap_ktype <- niche_overlapper(visit_cover, k.type.ss)

niche_overlap_plant.family <- niche_overlapper(visit_cover, plant.family)

niche_overlap <- bind_rows(no_plant.sp, 
                           no_ktype, 
                           no_plant.family)

# When bias is high, it could mean that there are few good options on which all (or many) BB species converge. But apparently it doesn't. When bias is high, BB spp are biased toward DIFFERENT flora, and thus their niches diverge. 
ggplot(bias, aes(dprime, mean.niche.overlap, color = criterion)) +
  geom_point(alpha = 0.1) +
  geom_smooth(method = "lm") +
  #geom_smooth(method = "gam", formula = y ~ s(x, k = 5), se = FALSE) +
  facet_wrap(~bb.sp) +
  ylim(c(0,1))

ggplot(bias, aes(dprime, max.niche.overlap, color = criterion)) +
  geom_point(alpha = 0.1) +
  geom_smooth(method = "lm") +
  #geom_smooth(method = "gam", formula = y ~ s(x, k = 5), se = FALSE) +
  facet_wrap(~bb.sp) +
  ylim(c(0,1))

ggplot(bias, aes(log(total.flower.cover), dprime, color = criterion)) +
  geom_point(alpha = 0.1) +
  geom_smooth(method = "lm") +
  #geom_smooth(method = "gam", formula = y ~ s(x, k = 5), se = FALSE) +
  facet_wrap(~bb.sp) +
  ylim(c(0,1))

ggplot(bias, aes(log(total.flower.cover), mean.niche.overlap, color = criterion)) +
  geom_point(alpha = 0.1) +
  geom_smooth(method = "lm") +
  #geom_smooth(method = "gam", formula = y ~ s(x, k = 5), se = FALSE) +
  facet_wrap(~bb.sp) +
  ylim(c(0,1))


brm.bias_niche <- brm(mean.niche.overlap ~ 1 + dprime + (1 | bb.sp),
                     cores = 3,
                     data = filter(bias, criterion == "ktype"))

parnames(brm.bias_niche)
conditional_effects(brm.bias_nich)
hypothesis(brm.bias_niche, "dprime = 0")
pp_check(brm.bias_niche)
```


